src/Firestore/WriteBatch.php

```diff
@@ -0,0 +1,536 @@
+     *
+     *     @type bool $merge If true, unwritten fields will be preserved.
+     *           Otherwise, they will be overwritten (removed). **Defaults to**
+     *           `false`.
+     * }
+     * @return WriteBatch
+     * @codingStandardsIgnoreEnd
+     */
+    public function set($documentName, array $fields, array $options = [])
+    {
+        $options += [
+            'merge' => false
+        ];
+
+        list($fields, $timestamps) = $this->valueMapper->findSentinels($fields);
+
+        if ($fields) {
+            $write = array_filter([
+                'fields' => $this->valueMapper->encodeValues($fields),
+                'updateMask' => $options['merge'] ? $this->valueMapper->encodeFieldPaths($fields) : null
+            ]);
+
+            $this->writes[] = $this->createDatabaseWrite(self::TYPE_UPDATE, $documentName, $write);
+        }
+
+        // Setting values to the server timestamp is implemented as a document tranformation.
+        $this->updateTransformations($documentName, $timestamps);
+
+        return $this;
+    }
+
+    /**
+     * Enqueue an update to a Firestore document.
+     *
+     * Merges provided data with data stored in Firestore.
+     *
+     * Calling this method on a non-existent document will raise an exception.
+     *
+     * This method supports various sentinel values, to perform special operations
+     * on fields. Available sentinel values are provided as methods, found in
+     * {@see Google\Cloud\Firestore\FieldValue}.
+     *
+     * To update a document using field paths, see
+     * {@see Google\Cloud\Firestore\WriteBatch::updatePaths()}.
+     *
+     * Example:
+     * ```
+     * $batch->update($documentName, [
+     *     'name' => 'John'
+     * ]);
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @param string $documentName The document to update.
+     * @param array $fields An array containing field names paired with their value.
+     * @param array $options Configuration options
+     * @return WriteBatch
+     * @codingStandardsIgnoreEnd
+     */
+    public function update($documentName, array $fields, array $options = [])
+    {
+        $options += [
+            'precondition' => ['exists' => true]
+        ];
+
+        list($fields, $timestamps, $deletes) = $this->valueMapper->findSentinels($fields);
+
+        // We only want to enqueue an update write if there are non-sentinel fields
+        // OR no timestamp sentinels are found.
+        // We MUST always enqueue at least one write, so if there are no fields
+        // and no timestamp sentinels, we can assume an empty write is intended
+        // and enqueue an empty UPDATE operation.
+        if ($fields || !$timestamps) {
+            $this->writes[] = $this->createDatabaseWrite(self::TYPE_UPDATE, $documentName, [
+                'fields' => $this->valueMapper->encodeValues($fields),
+                'updateMask' => array_merge($this->valueMapper->encodeFieldPaths($fields), $deletes)
```

Comment:
This should not use the deep-merge fields but the exact fields that the user provided: 'a.b' : { c { d: 'foo' }} should result in an update mask of 'a.b'.